File: ml_pipeline/ensemble_predictor.py
"""Ensemble Predictor for combining multiple sub-model predictions.

This module implements a weighted voting system that aggregates predictions
from three sub-models:
- Full-time Model (FT): Primary model with highest weight
- Half-time Model (HT): Intermediate predictions
- Pattern-based Model (PT): Pattern recognition model

Key features:
- Configurable weights (default: FT=0.5, HT=0.3, PT=0.2)
- Dynamic re-weighting when sub-models return null
- Conflict detection when top 2 outcomes have similar scores
- Deterministic output for reproducibility
"""

from typing import Dict, Optional, Tuple, List
import logging

logger = logging.getLogger(__name__)


class EnsemblePredictor:
    """Ensemble predictor implementing weighted voting logic."""
    
    # Default configuration weights
    DEFAULT_WEIGHTS = {
        "ft": 0.5,  # Full-time model
        "ht": 0.3,  # Half-time model
        "pt": 0.2,  # Pattern model
    }
    
    # Conflict threshold: if top 2 scores differ by less than this, flag as conflict
    CONFLICT_THRESHOLD = 0.1
    
    # Valid outcomes
    VALID_OUTCOMES = {"HOME", "DRAW", "AWAY", "home_win", "draw", "away_win"}
    
    def __init__(self, weights: Optional[Dict[str, float]] = None):
        """Initialize ensemble predictor with optional custom weights.
        
        Args:
            weights: Optional dict with keys 'ft', 'ht', 'pt' for model weights.
                    If not provided, uses DEFAULT_WEIGHTS.
        """
        self.weights = weights if weights else self.DEFAULT_WEIGHTS.copy()
        self._validate_weights()
    
    def _validate_weights(self) -> None:
        """Validate that weights are positive and sum close to 1.0."""
        required_keys = {"ft", "ht", "pt"}
        if not all(key in self.weights for key in required_keys):
            raise ValueError(f"Weights must contain keys: {required_keys}")
        
        if any(w < 0 for w in self.weights.values()):
            raise ValueError("All weights must be non-negative")
        
        total = sum(self.weights.values())
        if not (0.99 <= total <= 1.01):
            logger.warning(f"Weights sum to {total}, not 1.0. Will normalize during prediction.")
    
    def _normalize_outcome(self, outcome: str) -> str:
        """Normalize outcome to standard format (HOME, DRAW, AWAY)."""
        outcome_upper = outcome.upper()
        if outcome_upper in {"HOME", "HOME_WIN"}:
            return "HOME"
        elif outcome_upper == "DRAW":
            return "DRAW"
        elif outcome_upper in {"AWAY", "AWAY_WIN"}:
            return "AWAY"
        else:
            raise ValueError(f"Invalid outcome: {outcome}")
    
    def _normalize_outcome_for_output(self, outcome: str) -> str:
        """Normalize outcome to database format (home_win, draw, away_win)."""
        if outcome == "HOME":
            return "home_win"
        elif outcome == "DRAW":
            return "draw"
        elif outcome == "AWAY":
            return "away_win"
        return outcome
    
    def predict(
        self,
        full_time_prediction: Optional[str] = None,
        full_time_confidence: Optional[float] = None,
        half_time_prediction: Optional[str] = None,
        half_time_confidence: Optional[float] = None,
        pattern_prediction: Optional[str] = None,
        pattern_confidence: Optional[float] = None,
    ) -> Dict:
        """Calculate ensemble prediction from sub-model predictions.
        
        Args:
            full_time_prediction: FT model's predicted outcome (HOME/DRAW/AWAY or home_win/draw/away_win)
            full_time_confidence: FT model's confidence (0-1 range)
            half_time_prediction: HT model's predicted outcome
            half_time_confidence: HT model's confidence (0-1 range)
            pattern_prediction: PT model's predicted outcome
            pattern_confidence: PT model's confidence (0-1 range)
        
        Returns:
            Dict containing:
                - weights_used: Normalized weights used for each model
                - votes: Individual model predictions and confidences
                - scores: Aggregated scores for each outcome (HOME, DRAW, AWAY)
                - winner: Winning outcome
                - final_confidence: Confidence score of winning outcome
                - conflict_detected: Boolean indicating if top 2 outcomes are close
                - conflict_margin: Difference between top 2 scores
        
        Raises:
            ValueError: If all models return None or invalid inputs
        """
        # Collect valid models
        models = []
        if full_time_prediction is not None and full_time_confidence is not None:
            models.append(("full_time", full_time_prediction, full_time_confidence, "ft"))
        if half_time_prediction is not None and half_time_confidence is not None:
            models.append(("half_time", half_time_prediction, half_time_confidence, "ht"))
        if pattern_prediction is not None and pattern_confidence is not None:
            models.append(("pattern", pattern_prediction, pattern_confidence, "pt"))
        
        if not models:
            raise ValueError("At least one sub-model prediction must be provided")
        
        # Calculate total weight of active models
        total_weight = sum(self.weights[weight_key] for _, _, _, weight_key in models)
        
        if total_weight == 0:
            raise ValueError("Total weight of active models is zero")
        
        # Normalize weights for active models only
        normalized_weights = {}
        for _, _, _, weight_key in models:
            normalized_weights[weight_key] = self.weights[weight_key] / total_weight
        
        # Initialize scores for each outcome
        scores = {"HOME": 0.0, "DRAW": 0.0, "AWAY": 0.0}
        
        # Aggregate scores using weighted voting
        votes = {}
        for model_name, prediction, confidence, weight_key in models:
            # Validate confidence is in range
            if not (0.0 <= confidence <= 1.0):
                raise ValueError(f"{model_name} confidence must be in range [0, 1], got {confidence}")
            
            # Normalize outcome
            normalized_outcome = self._normalize_outcome(prediction)
            
            # Add weighted contribution to outcome score
            weight = normalized_weights[weight_key]
            scores[normalized_outcome] += confidence * weight
            
            # Record vote
            votes[model_name] = {
                "prediction": prediction,
                "confidence": confidence
            }
        
        # Determine winner (outcome with highest score)
        winner_outcome = max(scores.items(), key=lambda x: x[1])
        winner = winner_outcome[0]
        final_confidence = winner_outcome[1]
        
        # Check for conflict
        sorted_scores = sorted(scores.values(), reverse=True)
        conflict_margin = sorted_scores[0] - sorted_scores[1] if len(sorted_scores) > 1 else sorted_scores[0]
        conflict_detected = conflict_margin < self.CONFLICT_THRESHOLD
        
        # Build complete weights_used dict (including zeros for inactive models)
        weights_used = {
            "ft": normalized_weights.get("ft", 0.0),
            "ht": normalized_weights.get("ht", 0.0),
            "pt": normalized_weights.get("pt", 0.0)
        }
        
        # Build result
        result = {
            "weights_used": weights_used,
            "votes": votes,
            "scores": scores,
            "winner": self._normalize_outcome_for_output(winner),
            "final_confidence": round(final_confidence, 4),
            "conflict_detected": conflict_detected,
            "conflict_margin": round(conflict_margin, 4)
        }
        
        logger.info(
            f"Ensemble prediction: {result['winner']} "
            f"(confidence: {result['final_confidence']}, "
            f"conflict: {conflict_detected})"
        )
        
        return result
    
    def get_config(self) -> Dict[str, float]:
        """Get current weight configuration.
        
        Returns:
            Dict with model weights
        """
        return self.weights.copy()
    
    def update_config(self, new_weights: Dict[str, float]) -> None:
        """Update weight configuration.
        
        Args:
            new_weights: New weights dict with keys 'ft', 'ht', 'pt'
        
        Raises:
            ValueError: If weights are invalid
        """
        old_weights = self.weights.copy()
        self.weights = new_weights.copy()
        try:
            self._validate_weights()
            logger.info(f"Updated weights from {old_weights} to {new_weights}")
        except ValueError as e:
            self.weights = old_weights
            raise ValueError(f"Invalid weights: {e}")


def create_ensemble_predictor(weights: Optional[Dict[str, float]] = None) -> EnsemblePredictor:
    """Factory function to create an EnsemblePredictor instance.
    
    Args:
        weights: Optional custom weights dict
    
    Returns:
        EnsemblePredictor instance
    """
    return EnsemblePredictor(weights=weights)

File: 8888/utils/ensemblePredictor.ts
/**
 * Ensemble Predictor for combining multiple sub-model predictions.
 * 
 * This module implements a weighted voting system that aggregates predictions
 * from three sub-models:
 * - Full-time Model (FT): Primary model with highest weight
 * - Half-time Model (HT): Intermediate predictions
 * - Pattern-based Model (PT): Pattern recognition model
 * 
 * Key features:
 * - Configurable weights (default: FT=0.5, HT=0.3, PT=0.2)
 * - Dynamic re-weighting when sub-models return null
 * - Conflict detection when top 2 outcomes have similar scores
 * - Deterministic output for reproducibility
 */

export interface ModelWeights {
  ft: number;  // Full-time model weight
  ht: number;  // Half-time model weight
  pt: number;  // Pattern model weight
}

export interface ModelPrediction {
  prediction: string;
  confidence: number;
}

export interface EnsembleVotes {
  full_time?: ModelPrediction;
  half_time?: ModelPrediction;
  pattern?: ModelPrediction;
}

export interface OutcomeScores {
  HOME: number;
  DRAW: number;
  AWAY: number;
}

export interface EnsembleResult {
  weights_used: ModelWeights;
  votes: EnsembleVotes;
  scores: OutcomeScores;
  winner: string;
  final_confidence: number;
  conflict_detected: boolean;
  conflict_margin: number;
}

export interface EnsemblePredictionInput {
  full_time_prediction?: string | null;
  full_time_confidence?: number | null;
  half_time_prediction?: string | null;
  half_time_confidence?: number | null;
  pattern_prediction?: string | null;
  pattern_confidence?: number | null;
}

export class EnsemblePredictor {
  // Default configuration weights
  static readonly DEFAULT_WEIGHTS: ModelWeights = {
    ft: 0.5,  // Full-time model
    ht: 0.3,  // Half-time model
    pt: 0.2,  // Pattern model
  };
  
  // Conflict threshold: if top 2 scores differ by less than this, flag as conflict
  static readonly CONFLICT_THRESHOLD = 0.1;
  
  // Valid outcomes
  private static readonly VALID_OUTCOMES = new Set([
    'HOME', 'DRAW', 'AWAY', 'home_win', 'draw', 'away_win'
  ]);
  
  private weights: ModelWeights;
  
  constructor(weights?: Partial<ModelWeights>) {
    this.weights = {
      ...EnsemblePredictor.DEFAULT_WEIGHTS,
      ...weights
    };
    this.validateWeights();
  }
  
  private validateWeights(): void {
    const { ft, ht, pt } = this.weights;
    
    if (ft < 0 || ht < 0 || pt < 0) {
      throw new Error('All weights must be non-negative');
    }
    
    const total = ft + ht + pt;
    if (total < 0.99 || total > 1.01) {
      console.warn(`Weights sum to ${total}, not 1.0. Will normalize during prediction.`);
    }
  }
  
  private normalizeOutcome(outcome: string): 'HOME' | 'DRAW' | 'AWAY' {
    const outcomeUpper = outcome.toUpperCase();
    if (outcomeUpper === 'HOME' || outcomeUpper === 'HOME_WIN') {
      return 'HOME';
    } else if (outcomeUpper === 'DRAW') {
      return 'DRAW';
    } else if (outcomeUpper === 'AWAY' || outcomeUpper === 'AWAY_WIN') {
      return 'AWAY';
    } else {
      throw new Error(`Invalid outcome: ${outcome}`);
    }
  }
  
  private normalizeOutcomeForOutput(outcome: 'HOME' | 'DRAW' | 'AWAY'): string {
    if (outcome === 'HOME') {
      return 'home_win';
    } else if (outcome === 'DRAW') {
      return 'draw';
    } else if (outcome === 'AWAY') {
      return 'away_win';
    }
    return outcome;
  }
  
  predict(input: EnsemblePredictionInput): EnsembleResult {
    const {
      full_time_prediction,
      full_time_confidence,
      half_time_prediction,
      half_time_confidence,
      pattern_prediction,
      pattern_confidence
    } = input;
    
    // Collect valid models
    interface ModelInfo {
      name: 'full_time' | 'half_time' | 'pattern';
      prediction: string;
      confidence: number;
      weightKey: 'ft' | 'ht' | 'pt';
    }
    
    const models: ModelInfo[] = [];
    
    if (full_time_prediction != null && full_time_confidence != null) {
      models.push({
        name: 'full_time',
        prediction: full_time_prediction,
        confidence: full_time_confidence,
        weightKey: 'ft'
      });
    }
    
    if (half_time_prediction != null && half_time_confidence != null) {
      models.push({
        name: 'half_time',
        prediction: half_time_prediction,
        confidence: half_time_confidence,
        weightKey: 'ht'
      });
    }
    
    if (pattern_prediction != null && pattern_confidence != null) {
      models.push({
        name: 'pattern',
        prediction: pattern_prediction,
        confidence: pattern_confidence,
        weightKey: 'pt'
      });
    }
    
    if (models.length === 0) {
      throw new Error('At least one sub-model prediction must be provided');
    }
    
    // Calculate total weight of active models
    const totalWeight = models.reduce((sum, model) => sum + this.weights[model.weightKey], 0);
    
    if (totalWeight === 0) {
      throw new Error('Total weight of active models is zero');
    }
    
    // Normalize weights for active models only
    const normalizedWeights: Partial<ModelWeights> = {};
    for (const model of models) {
      normalizedWeights[model.weightKey] = this.weights[model.weightKey] / totalWeight;
    }
    
    // Initialize scores for each outcome
    const scores: OutcomeScores = { HOME: 0, DRAW: 0, AWAY: 0 };
    
    // Aggregate scores using weighted voting
    const votes: EnsembleVotes = {};
    
    for (const model of models) {
      // Validate confidence is in range
      if (model.confidence < 0 || model.confidence > 1) {
        throw new Error(
          `${model.name} confidence must be in range [0, 1], got ${model.confidence}`
        );
      }
      
      // Normalize outcome
      const normalizedOutcome = this.normalizeOutcome(model.prediction);
      
      // Add weighted contribution to outcome score
      const weight = normalizedWeights[model.weightKey]!;
      scores[normalizedOutcome] += model.confidence * weight;
      
      // Record vote
      votes[model.name] = {
        prediction: model.prediction,
        confidence: model.confidence
      };
    }
    
    // Determine winner (outcome with highest score)
    const entries = Object.entries(scores) as [keyof OutcomeScores, number][];
    const [winner, finalConfidence] = entries.reduce((max, entry) => 
      entry[1] > max[1] ? entry : max
    );
    
    // Check for conflict
    const sortedScores = Object.values(scores).sort((a, b) => b - a);
    const conflictMargin = sortedScores.length > 1 
      ? sortedScores[0] - sortedScores[1] 
      : sortedScores[0];
    const conflictDetected = conflictMargin < EnsemblePredictor.CONFLICT_THRESHOLD;
    
    // Build complete weights_used dict (including zeros for inactive models)
    const weightsUsed: ModelWeights = {
      ft: normalizedWeights.ft ?? 0,
      ht: normalizedWeights.ht ?? 0,
      pt: normalizedWeights.pt ?? 0
    };
    
    // Build result
    const result: EnsembleResult = {
      weights_used: weightsUsed,
      votes,
      scores,
      winner: this.normalizeOutcomeForOutput(winner),
      final_confidence: Math.round(finalConfidence * 10000) / 10000, // Round to 4 decimals
      conflict_detected: conflictDetected,
      conflict_margin: Math.round(conflictMargin * 10000) / 10000
    };
    
    console.log(
      `Ensemble prediction: ${result.winner} ` +
      `(confidence: ${result.final_confidence}, conflict: ${conflictDetected})`
    );
    
    return result;
  }
  
  getConfig(): ModelWeights {
    return { ...this.weights };
  }
  
  updateConfig(newWeights: Partial<ModelWeights>): void {
    const oldWeights = { ...this.weights };
    this.weights = {
      ...this.weights,
      ...newWeights
    };
    
    try {
      this.validateWeights();
      console.log(`Updated weights from`, oldWeights, `to`, this.weights);
    } catch (error) {
      this.weights = oldWeights;
      throw new Error(`Invalid weights: ${error instanceof Error ? error.message : error}`);
    }
  }
}

/**
 * Factory function to create an EnsemblePredictor instance.
 */
export function createEnsemblePredictor(weights?: Partial<ModelWeights>): EnsemblePredictor {
  return new EnsemblePredictor(weights);
}
